### ISSUE TITLE
1.10.33.1-beta clang compile errors

### ISSUE THREAD

**Issue by centminmod** [pagespeed/ngx_pagespeed#1079]
Attempting to compile Nginx 1.9.9 via clang 3.4.2 on centos 6.7 and get following errors

    make -f objs/Makefile install
    make[1]: Entering directory `/svr-setup/nginx-1.9.9'
    ccache /usr/bin/clang -ferror-limit=0 -c -I/usr/local/include/luajit-2.1  -pipe  -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Werror -g -m64 -mtune=native -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wno-sign-compare -Wno-string-plus-int -Wno-deprecated-declarations -Wno-unused-parameter -Wno-unused-const-variable -Wno-conditional-uninitialized -Wno-mismatched-tags -Wno-c++11-extensions -Wno-sometimes-uninitialized -Wno-parentheses-equality -Wno-tautological-compare -Wno-self-assign -Wno-deprecated-register -Wno-deprecated -Wno-invalid-source-encoding -Wno-pointer-sign -Wno-parentheses -Wno-enum-conversion  -D_GLIBCXX_USE_CXX11_ABI=0 -DNDK_SET_VAR -DNDK_UPSTREAM_LIST -DNDK_SET_VAR --std=c++11  -I src/core -I src/event -I src/event/modules -I src/os/unix -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/chromium/src -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/google-sparsehash/src -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/protobuf/src -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/re2/src -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/out/Release/obj/gen -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/out/Release/obj/gen/protoc_out/instaweb -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/apr/src/include -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/aprutil/src/include -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/apr/gen/arch/linux/x64/include -I ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/third_party/aprutil/gen/arch/linux/x64/include -I ../ngx_devel_kit-0.2.19/objs -I objs/addon/ndk -I /usr/local/include/luajit-2.1 -I ../lua-nginx-module-0.9.19/src/api -I ../nginx_upstream_check_module-0.3.0 -I ../pcre-8.38 -I ../libressl-2.2.5/.openssl/include -I objs -I src/http -I src/http/modules -I src/http/v2 -I ..    /ngx_devel_kit-0.2.19/src -I src/mail -I src/stream \
                    -o objs/addon/src/log_message_handler.o \
                    ../ngx_pagespeed-release-1.10.33.1-beta/src/log_message_handler.cc
    In file included from ../ngx_pagespeed-release-1.10.33.1-beta/src/log_message_handler.cc:26:
    In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/string:42:
    In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/char_traits.h:41:
    In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/stl_algobase.h:66:
    In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/stl_pair.h:60:
    /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/move.h:57:14: error: rvalue reference to type 'basic_string<[3 * ...]>' cannot bind to lvalue of type 'basic_string<[3 * ...]>'
        { return __t; }
                 ^~~
    /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/stl_vector.h:747:27: note: in instantiation of function template specialization 'std::move<std::basic_string<char> &>' requested here
          { emplace_back(std::move(__x)); }
                              ^
    ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/pagespeed/kernel/base/string_util.h:688:6: note: in instantiation of member function 'std::vector<std::basic_string<char>, std::allocator<std::basic_string<char> > >    ::push_back' requested here
      v->push_back(GoogleString());
         ^
    In file included from ../ngx_pagespeed-release-1.10.33.1-beta/src/log_message_handler.cc:32:
    In file included from ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/pagespeed/kernel/base/string_util.h:26:
    In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/vector:69:
    /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/vector.tcc:96:9: error: no matching function for call to 'forward'
                                        std::forward<_Args>(__args)...);
                                        ^~~~~~~~~~~~~~~~~~~
    /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/stl_vector.h:747:9: note: in instantiation of function template specialization 'std::vector<std::basic_string<char>, std::allocator<    std::basic_string<char> > >::emplace_back<std::basic_string<char> >' requested here
          { emplace_back(std::move(__x)); }
            ^
    ../ngx_pagespeed-release-1.10.33.1-beta/psol/include/pagespeed/kernel/base/string_util.h:688:6: note: in instantiation of member function 'std::vector<std::basic_string<char>, std::allocator<std::basic_string<char> > >    ::push_back' requested here
      v->push_back(GoogleString());
         ^
    /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/move.h:51:5: note: candidate function [with _Tp = std::basic_string<char>] not viable: no known conversion from 'std::basic_string<char>' to     'typename std::identity<basic_string<char> >::type &&' (aka 'std::basic_string<char> &&') for 1st argument
        forward(typename std::identity<_Tp>::type&& __t)
        ^
    2 errors generated.
    make[1]: *** [objs/addon/src/log_message_handler.o] Error 1
    make[1]: Leaving directory `/svr-setup/nginx-1.9.9'
    make: *** [install] Error 2

this will be problematic as 90% of centmin mod nginx users use CentOS 6.7 still

using clang because it's much faster compiler and lower memory footprint

would be awesome to have a clang friendly version of ngx_pagespeed :)

**Comment by jeffkaufman** [comment]
@crowell were you looking at something like this?

**Comment by crowell** [comment]
i was looking at building mod_pagespeed/psol with clang, not sure if this
is related yet, but I'll take a look

On Fri, Dec 18, 2015 at 8:46 PM, Jeff Kaufman <notifications@github.com>
wrote:

> @crowell <https://github.com/crowell> were you looking at something like
> this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pagespeed/ngx_pagespeed/issues/1079#issuecomment-165934612>
> .
>



-- 
Jeff Crowell
https://github.com/crowell

**Comment by centminmod** [comment]
@crowell if i switch back to using GCC via devtoolset-3's GCC 4.9.1 it compiles fine https://community.centminmod.com/posts/22656/. Just when using Clang 3.4.2 the above errors occur. HTH

**Comment by crowell** [comment]
I can't reproduce on ubuntu 14.04 with clang version 

```
jcrowell@jcrowell0:~/nginx-1.8.0$ clang --version
Ubuntu clang version 3.4-1ubuntu3 (tags/RELEASE_34/final) (based on LLVM 3.4)
Target: x86_64-pc-linux-gnu
Thread model: posix
```

i'll take a look on centos6.7 now.

**Comment by crowell** [comment]
@centminmod by default clang will use the system gcc's includes. can you try setting `CC=--gcc-toolchain=/opt/rh/devtoolset-2/root/usr/` (or other appropriate path) after installing the devtoolset

**Comment by centminmod** [comment]
@crowell you're a lifesaver ! workaround hack worked wonderfully on centos 6.7 ! :+1: 

nginx -V
nginx version: nginx/1.9.9
built by clang 3.4.2 (tags/RELEASE_34/dot2-final)
built with LibreSSL 2.2.5
TLS SNI support enabled
configure arguments: --with-ld-opt='-lrt -ljemalloc -Wl,-z,relro -Wl,-rpath,/usr/local/lib' --with-cc-opt='-m64 -mtune=native -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wno-sign-compare -Wno-string-plus-int -Wno-deprecated-declarations -Wno-unused-parameter -Wno-unused-const-variable -Wno-conditional-uninitialized -Wno-mismatched-tags -Wno-c++11-extensions -Wno-sometimes-uninitialized -Wno-parentheses-equality -Wno-tautological-compare -Wno-self-assign -Wno-deprecated-register -Wno-deprecated -Wno-invalid-source-encoding -Wno-pointer-sign -Wno-parentheses -Wno-enum-conversion' --sbin-path=/usr/local/sbin/nginx --conf-path=/usr/local/nginx/conf/nginx.conf --with-http_stub_status_module --with-http_secure_link_module --with-openssl-opt=enable-tlsext --add-module=../nginx-module-vts --with-libatomic --with-threads --with-stream --with-stream_ssl_module --with-http_gzip_static_module --add-module=../ngx_pagespeed-release-1.10.33.1-beta --with-http_sub_module --with-http_addition_module --with-http_image_filter_module --with-http_geoip_module --with-http_realip_module --add-module=../nginx-accesskey-2.0.3 --add-module=../nginx-http-concat-master --add-module=../ngx-fancyindex-ngx-fancyindex --add-module=../ngx_cache_purge-2.3 --add-module=../ngx_devel_kit-0.2.19 --add-module=../set-misc-nginx-module-0.29 --add-module=../echo-nginx-module-0.58 --add-module=../redis2-nginx-module-0.12 --add-module=../ngx_http_redis-0.3.7 --add-module=../lua-nginx-module-0.9.19 --add-module=../lua-upstream-nginx-module-0.04 --add-module=../lua-upstream-cache-nginx-module-0.1.1 --add-module=../nginx_upstream_check_module-0.3.0 --add-module=../openresty-memc-nginx-module-4f6f78f --add-module=../openresty-srcache-nginx-module-ffa9ab7 --add-module=../headers-more-nginx-module-0.28 --with-pcre=../pcre-8.38 --with-pcre-jit --with-http_ssl_module --with-http_v2_module --with-openssl=../libressl-2.2.5

**Comment by crowell** [comment]
awesome!

@centminmod the "real" solution is to build like this now.

```bash
export CC=clang
export CFLAGS="--gcc-toolchain=/opt/rh/devtoolset-2/root/usr/"
./configure --add-module=$HOME/ngx_pagespeed-release-${NPS_VERSION}-beta "--with-ld-opt=--gcc-toolchain=/opt/rh/devtoolset-2/root/
usr/" # or whatever other configure options you add normally
make
```

I'll make sure to doc this too.

**Comment by centminmod** [comment]
thanks @crowell can confirm exporting CFLAGS instructions you outlined work well too tested on newly released 1.10.33.2-beta :)

```
nginx -V
nginx version: nginx/1.9.9
built with LibreSSL 2.2.5
TLS SNI support enabled
configure arguments: --with-ld-opt='-lrt -ljemalloc -Wl,-z,relro -Wl,-rpath,/usr/local/lib --gcc-toolchain=/opt/rh/devtoolset-3/root/usr/' --with-cc-opt='-m64 -mtune=native -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wno-sign-compare -Wno-string-plus-int -Wno-deprecated-declarations -Wno-unused-parameter -Wno-unused-const-variable -Wno-conditional-uninitialized -Wno-mismatched-tags -Wno-c++11-extensions -Wno-sometimes-uninitialized -Wno-parentheses-equality -Wno-tautological-compare -Wno-self-assign -Wno-deprecated-register -Wno-deprecated -Wno-invalid-source-encoding -Wno-pointer-sign -Wno-parentheses -Wno-enum-conversion' --sbin-path=/usr/local/sbin/nginx --conf-path=/usr/local/nginx/conf/nginx.conf --with-http_stub_status_module --with-http_secure_link_module --with-openssl-opt=enable-tlsext --add-module=../nginx-module-vts --with-libatomic --with-threads --with-stream --with-stream_ssl_module --with-http_gzip_static_module --add-module=../ngx_pagespeed-release-1.10.33.2-beta --with-http_sub_module --with-http_addition_module --with-http_image_filter_module --with-http_geoip_module --with-http_realip_module --add-module=../nginx-accesskey-2.0.3 --add-module=../nginx-http-concat-master --add-module=../ngx-fancyindex-ngx-fancyindex --add-module=../ngx_cache_purge-2.3 --add-module=../ngx_devel_kit-0.2.19 --add-module=../set-misc-nginx-module-0.29 --add-module=../echo-nginx-module-0.58 --add-module=../redis2-nginx-module-0.12 --add-module=../ngx_http_redis-0.3.7 --add-module=../lua-nginx-module-0.9.19 --add-module=../lua-upstream-nginx-module-0.04 --add-module=../lua-upstream-cache-nginx-module-0.1.1 --add-module=../nginx_upstream_check_module-0.3.0 --add-module=../openresty-memc-nginx-module-4f6f78f --add-module=../openresty-srcache-nginx-module-ffa9ab7 --add-module=../headers-more-nginx-module-0.28 --with-pcre=../pcre-8.38 --with-pcre-jit --with-http_ssl_module --with-http_v2_module --with-openssl=../libressl-2.2.5
```

**Comment by morlovich** [comment]
So the trickiness here --- and why it works with the similar version for Jeff on Ubuntu 14 --- is that clang on Linux uses the system libstdc++ provided by gcc --- so if your gcc is old, you end up with a modern C++-11 compiler trying to build an ancient version of the library that's not quite up to the spec. Oh, and of course there are tons of combinations. There are actually some patches around to make different combinations work, but they change system libstdc++ headers, and so are probably a Very Bad Idea (tm). The suggested workaround (which just uses a separate newer gcc) is much safer.
